dll/win32/browseui/aclistisf.cpp

```diff
@@ -20,83 +21,411 @@
+        case LT_MYCOMPUTER:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_MYCOMPUTER)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_DRIVES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_MYCOMPUTER\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_FAVORITES:
+            ++m_iNextLocation;
+            if (m_dwOptions & ACLO_FAVORITES)
+            {
+                CComHeapPtr<ITEMIDLIST> pidl;
+                hr = SHGetSpecialFolderLocation(NULL, CSIDL_FAVORITES, &pidl);
+                if (FAILED(hr))
+                    return S_FALSE;
+                hr = SetLocation(pidl);
+                if (SUCCEEDED(hr))
+                {
+                    TRACE("LT_FAVORITES\n");
+                    return hr;
+                }
+            }
+            // FALL THROUGH
+        case LT_MAX:
+        default:
+            TRACE("LT_MAX\n");
+            return S_FALSE;
+    }
+}
+
+HRESULT CACListISF::SetLocation(LPCITEMIDLIST pidl)
+{
+    TRACE("SetLocation(%p, %p)\n", this, pidl);
+
+    m_pEnumIDList.Release();
+    m_pShellFolder.Release();
+    m_pidlLocation.Free();
+
+    if (!pidl)
+        return E_FAIL;
+
+    LPITEMIDLIST pidlClone = ILClone(pidl);
+    if (!pidlClone)
+    {
+        ERR("Out of memory\n");
+        return E_OUTOFMEMORY;
+    }
+    m_pidlLocation.Attach(pidlClone);
+
+    CComPtr<IShellFolder> pFolder;
+    HRESULT hr = SHGetDesktopFolder(&pFolder);
+    if (FAILED(hr))
+    {
+        ERR("SHGetDesktopFolder failed: 0x%lX\n", hr);
+        return hr;
+    }
+
+    if (!ILIsEmpty(pidl))
+        pFolder->BindToObject(pidl, NULL, IID_PPV_ARG(IShellFolder, &m_pShellFolder));
+    else
+        m_pShellFolder.Attach(pFolder.Detach());
+
+    SHCONTF Flags = SHCONTF_FOLDERS | SHCONTF_INIT_ON_FIRST_NEXT;
+    if (m_fShowHidden)
+        Flags |= SHCONTF_INCLUDEHIDDEN;
+    if (!(m_dwOptions & ACLO_FILESYSDIRS))
+        Flags |= SHCONTF_NONFOLDERS;
+
+    hr = m_pShellFolder->EnumObjects(NULL, Flags, &m_pEnumIDList);
+    if (hr != S_OK)
+    {
+        ERR("EnumObjects failed: 0x%lX\n", hr);
+        hr = E_FAIL;
+        return hr;
+    }
+    return hr;
+}
+
+HRESULT CACListISF::GetDisplayName(LPCITEMIDLIST pidlChild, CComHeapPtr<WCHAR>& pszChild)
+{
+    TRACE("GetDisplayName(%p, %p)\n", this, pidlChild);
+    pszChild.Free();
+
+    STRRET StrRet;
+    DWORD dwFlags = SHGDN_INFOLDER | SHGDN_FORPARSING | SHGDN_FORADDRESSBAR;
+    HRESULT hr = m_pShellFolder->GetDisplayNameOf(pidlChild, dwFlags, &StrRet);
+    if (FAILED(hr))
+    {
+        dwFlags = SHGDN_INFOLDER | SHGDN_FORPARSING;
+        hr = m_pShellFolder->GetDisplayNameOf(pidlChild, dwFlags, &StrRet);
+        if (FAILED(hr))
```

Comment:
FAILED_UNEXPECTEDLY?
Same for all cases that are not expected to fail.

Reply:
OK.